ExecHashMap.hpp

```diff
@@ -0,0 +1,110 @@
+#include <iostream>
+#include <iterator>
+#include <numeric>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <nlohmann/json.hpp>
+
+typedef std::vector<std::string> IOVector;
+typedef std::pair<IOVector, IOVector> IOPair;
+
+/**
+ * @brief A hash map that maps inputs to a vector of outputs
+ * @details Key is inputs (a vector), value is a vector of outputs
+ * (a vector of vectors) of multiple executions
+ */
+class ExecHashMap {
+private:
+  struct VectorHasher {
+    int operator()(const std::vector<std::string> &V) const {
+      std::string all = std::accumulate(
+          V.begin(), V.end(), std::string(),
+          [&](std::string x, std::string y) { return x + "," + y; });
+      return std::hash<std::string>{}(all);
+    }
+  };
+
+  std::unordered_map<IOVector, std::vector<IOVector>, VectorHasher> map;
+  int value_capacity;
+
+public:
+  ExecHashMap() : ExecHashMap(0) {}
+
+  /**
+   * @brief Construct a new Exec Hash Map object
+   * @param cap the capacity of the value vector (maxmium length)
+   */
+  ExecHashMap(int cap) : value_capacity(cap) {}
+
+  /**
+   * @brief Insert a pair of inputs and outputs to the hash map
+   * @param io: a pair of inputs (vector) and outputs (vector)
+   */
+  void insert(const IOPair &io) {
+    IOVector inputs = io.first;
+    IOVector outputs = io.second;
+    if (map[inputs].size() < value_capacity) {
+      map[inputs].push_back(outputs);
+    }
+  }
+
+  /**
+   * @brief Get the outputs of a given input
+   * @param xs: inputs
+   * @return a vector of outputs
+   */
+  std::vector<IOVector> &operator[](const IOVector &xs) { return map[xs]; }
+
+  int size() { return map.size(); }
+
+  nlohmann::json to_json() const {
+    nlohmann::json j;
+    for (auto &kv : map) {
+      j += nlohmann::json{{kv.first, kv.second}};
+    }
+    return j;
+  }
+};
+
+class ReportTable {
+private:
+  std::unordered_map<std::string, ExecHashMap> table;
+  int value_capacity;
+
+public:
+  ReportTable() {}
+
+  /**
+   * @brief Construct a new Report Table object
+   * @param cap the capacity of the value vector and report table
+   */
+  ReportTable(int cap) : value_capacity(cap) {}
```

Comment:
should we use different values for the capacity of ReporTable (maximum number of functions to be reported) and capacity of ExecHashMap (maximum number of outputs vectors corresponding to one vector of inputs)